Write "getting started" documentation (#230)
lab-cosmo/equistore: docs/src/get-started/concepts.rst

.. _userdoc-core-concepts:

Core concepts
=============

This page explains the main types and core concepts you need to understand to

> use equistore, without tying it to how a specific library is using equistore to
> store it's own data.

Reviewer: ```suggestion
use equistore. We discuss equistore without the context
of any specific library, but provide as example, but provide 
an illustrative example later with SOAP to give further context.
```

While staying general at first is great, people may really benefit from one concrete demonstration.

Author: Agree with that, but I would put the example in a separate page (with a link to it from this one)

Reviewer: Agreed!

.. py:currentmodule:: equistore

.. _concept-TensorMap:

TensorMap
^^^^^^^^^

The core type of equistore is the :py:class:`TensorMap`: a high dimensional
block-sparse tensor containing both data and metadata. A TensorMap contains a

> list of block (represented as :ref:`concept-TensorBlock`), each associated with

Reviewer: ```suggestion
list of blocks (represented as :ref:`concept-TensorBlock`), each associated with
```

a key. The set of all keys is stored in a :ref:`concept-Labels` object
(described below), which can be accessed in Python via
:py:attr:`TensorMap.keys`.

The keys can contain multiple dimensions (in the illustration below we have two
dimensions named ``lambda`` and ``sigma``), and each entry in the keys has one
integer value for each dimension. Here for example, the first block is
associated with ``lambda = 0`` and ``sigma = 0``, while the second block is
associated with ``lambda = 0`` and ``sigma = 1``, and so on.

.. figure:: ../../static/images/TensorMap.*
    :width: 600px
    :align: center

    Illustration of an equistore TensorMap object, made of a set of keys and
    associated :ref:`concept-TensorBlock`.

Different key dimensions can have different purposes, but some typical keys
dimensions you'll encounter when working with atomistic data are the following:

> - **atomic species dimensions**: when using a one-hot encoding of different
>   atomic species (or atomic elements) in a structure, the resulting data is
>   sparse, containing implicit zeros if a given species is not present in a
>   structure. This is the case of the ``species_center`` and various
>   ``species_neighbor`` key dimensions produced by `rascaline`_.

Reviewer: ```suggestion
- **atomic species**: for different elements, we often encode their
  using one-hot encodings. However, when a species is not present
  in a dataset, traditional numpy-like formats will be sparse, with 
  implicit zeros wherever the species is not present.
  This is the case of the ``species_center`` and various
  ``species_neighbor`` key dimensions produced by `rascaline`_.
```

Author: > However, when a species is not present in a dataset, traditional numpy-like formats will be sparse, with implicit zeros wherever the species is not present.

that's not the case: numpy-like storage would have to store **explicit** zeros

- **symmetry markers**: Another use case for equistore is to store and
  manipulate equivariant data, i.e. data that transforms in a known, specific
  way when the corresponding atomic structure is transformed. This is typically
  used to represent the symmetry property of some data with respect to
  rotations, by decomposing the properties of interest on a basis of spherical
  harmonics. When handling this kind of data, it is convenient to store and
  manipulate the data corresponding to different spherical harmonics (or
  generally different irreducible representations of the symmetry group)
  separately. This is the case of the ``spherical_harmonics_l`` key dimension
  produced by `rascaline`_: different blocks will contain the :math:`\ell = 1`
  and :math:`\ell = 2` parts of an equivariant representation.

.. _rascaline: https://github.com/Luthaf/rascaline/

.. _concept-Labels:

Labels
^^^^^^

A fundamental part of equistore is to carry simultaneously the data used in

> machine learning and the associated metadata. The first kind of metadata we
> encountered was the keys of a :py:class:`TensorMap`, stored as an instance of
> the :py:class:`Labels` class. This class is also used to store all other
> metadata in equistore, i.e. all the metadata associated with a given
> :py:class:`TensorBlock`.

Reviewer: ```suggestion
machine learning and the associated metadata. :py:class:`Labels` 
serves as a superclass for many forms of metadata, including
our :py:class:`TensorMap` keys.
```

Author: Labels is not a superclass (i.e. parent class w.r.t. inheritance), it is the class/type which all metadata is an instance of

.. figure:: ../../static/images/Labels.*
    :width: 600px
    :align: center

    Illustration of two different :py:class:`Labels` instances, corresponding to
    potential *samples* (green, on the left) and *properties* (red, on the
    right) of a :py:class:`TensorBlock`.

A set of :py:class:`Labels` can be seen as a two dimensional array of integers,
where each row corresponds to an entry in the data, and each column is a
*dimension*, which is named. For example, in the illustration above, the set of
Labels on the left has two dimensions (``structure`` and ``center``), and 10
entries (10 rows); while the Labels on the right has four dimensions and 8
entries. Depending on the language you use, :py:class:`Labels` entries and
dimensions' names can be accessed and manipulated in different ways, please
refer to the corresponding :ref:`API documentation <api-reference>` for more
information.

.. _concept-TensorBlock:

TensorBlock
^^^^^^^^^^^

The final core object of equistore is the :py:class:`TensorBlock`, containing a
dense array of data and metadata describing the different axes of this array.
The simplest possible TensorBlock is represented below, and contains three things:

> - a 2-dimensional data array;
> - metadata describing the rows of this array, called **samples** and stored as a
>   set of :py:class:`Labels`;
> - metadata describing the columns of this array, called **properties**, also
>   stored as a set of :py:class:`Labels`.

Reviewer: ```suggestion
- `Data`: a 2-dimensional data array;
- `Samples`: metadata describing the rows of this array, called **samples** and stored as a
  set of :py:class:`Labels`;
- `Properties`: metadata describing the columns of this array, called **properties**, also
  stored as a set of :py:class:`Labels`.
```

The samples store information about **what objects** does the data represent,
while properties store information about **how** these objects are represented.
Taking a couple of examples for clarity:

- if a TensorBlock is storing an atom-centered machine learning representation,
  the samples would contain the index of the atom ``"center"`` and the index of
  the corresponding ``"structure"``; while the properties would contain
  information about the e.g. the basis functions used to define the
  representation. The :ref:`Labels figure <fig-labels>` above contains an
  example of samples and properties that one would find in machine learning

>   representation.

Reviewer: Can you start with a simpler example that this? It's a bit "being thrown into the deep end" on a complex thought

- Alternatively, for a TensorBlock storing the energy of a list of structures,
  the samples would contain only a single ``"structure"`` index, and the
  properties would contain a single ``"energy"`` dimension with a single entry.

In general, for a 2-dimensional data array, the value at index ``(i, j)`` is

> described by the ``i``'th entry of the samples and the ``j``'th entry of the

Reviewer: This renders weird

    - `PyTorch's Tensor`_ from Python, including full support for autograd and
      different device (data living on CPU memory, GPU memory, …),
    - `Rust's ndarray`_ from Rust, more specifically ``ndarray::ArrayD<f64>``,
    - A very bare-bone N-dimensional array in equistore C++ API:
      :cpp:class:`equistore::SimpleDataArray`

    It is possible to integrate new array types with equistore, look into the
    :py:func:`equistore.data.register_external_data_wrapper` function in Python, the
    :c:struct:`eqs_array_t` struct in C, the :cpp:class:`equistore::DataArrayBase`
    abstract base class in C++, and the `equistore::Array`_ trait in Rust.

.. _Numpy's ndarray: https://numpy.org/doc/stable/reference/arrays.ndarray.html
.. _PyTorch's Tensor: https://pytorch.org/docs/stable/tensors.html
.. _Rust's ndarray: https://docs.rs/ndarray/
.. _equistore::Array: ../reference/rust/equistore/trait.Array.html

Gradients
---------

In addition to storing data and metadata together, a :py:class:`TensorBlock` can
also store values and gradients together. The gradients are stored in another
:py:class:`TensorBlock`, associated with a **parameter** name, describing with

> respect to **what** the gradients are taken. Regarding metadta, the gradient

Reviewer: ```suggestion
respect to **what** the gradients are taken. Regarding metadata, the gradient
```

handling spherical harmonics (which are the irreducible representation of the
`group of 3D rotations`_ :math:`SO(3)`), all the spherical harmonics with the same
:math:`\ell` and corresponding :math:`m` should be considered simultaneously:
the different :math:`m` are **components** of a single irreducible
representation.

.. _group of 3D rotations: https://en.wikipedia.org/wiki/3D_rotation_group

When handling the gradients of equivariant data, we quickly realize that we
might need more than one component in a given :py:class:`TensorBlock`. Gradients
with respect to positions of an equivariant representation based on spherical
harmonics will need both the gradient direction :math:`x/y/z` and the spherical
harmonics :math:`m` as components. This has two implications:

- :py:class:`TensorBlock` can have an arbitrary number of components, which will
  always occur *"in between"* samples and properties metadata;
- when values in a :py:class:`TensorBlock` have components, and gradient with
  respect to some parameter would add more components, the resulting gradient
  components will contain first the new, gradient-specific components, and later

>   all of the components already present in the values.

Reviewer: I would spin this into a new page for `What's in a Tensor Block?`, as these are as core of concepts like `TensorMap`, `Labels`, and `TensorBlock`, and keeping the page shorter will help with comprehension.
